core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -734,9 +734,7 @@ uint32_t Cuda::impl_instance_id() const noexcept {
   return m_space_instance->impl_get_instance_id();
 }
 
-cudaStream_t Cuda::cuda_stream() const {
-  return m_space_instance->get_stream();
-}
+cudaStream_t Cuda::cuda_stream() const { return m_space_instance->m_stream; }
```

Comment:
Isn't this a very important change ? From https://github.com/kokkos/kokkos/blob/79d2a342fed1daa31508f50d4c3b15d4d61c82f1/core/src/Cuda/Kokkos_Cuda_Instance.hpp#L189-L194, it seems that this function was "setting the device and returning the stream", and your change implies that it now does "return the stream", nothing else.

Reply:
I think it's useful to set the device here since it's user-visible.

Reply:
(addressing @romintomasetti comment) I think that is a reasonable change, and considering that the multi GPUs support is experimental, we are free to do so w/o guarding.

> I think it's useful to set the device here since it's user-visible.

Not sure what you are saying.

Reply:
If users are asking for `cuda_stream()`, they want to use it for something, and it's convenient if they don't have to take care of setting the correct device (which we assume to be rather cheap and easy to forget). It's not clear to me that it's better to change the behavior now. All the other changes are internal.

Reply:
I think it would then be easy for the user to assume that they can use the stream without worrying setting the device, when in fact they can only use the stream right after getting it without setting the device, but if any other device is set during the life of the stream they won't be safe.

Reply:
My comment wasn't about reverting this change.

I think this change goes into the right direction.

If a user asks for the underlying `Cuda` stream, we should consider that this user is advanced, and we should do nothing else than returning the `Cuda` stream. IMO, it's the user's responsibility to care about the current device when he/she is dealing with pure native things.

> it's convenient if they don't have to take care of setting the correct device

Moreover, there exists operations that you can perform on a stream that do no necessitate that the current device is set correctly (*e.g.* `cudaStreamSynchronize`, as discussed in #7562).